3_visualize/src/summarize_nhdv2_attr.R

```diff
@@ -2,28 +2,26 @@ summarize_nhdv2_attr <- function(attr_data,fileout){
   #' 
   #' @description This function summarizes each of the downloaded NHDv2 attribute variables across all river segments within the network
   #'
-  #' @param attr_data list containing the processed NHDv2 attribute data; data frames must include column "PRMS_segid" 
+  #' @param attr_data data frame containing the processed NHDv2 attribute data;
+  #' must include column "PRMS_segid"
   #' @param fileout a character string that indicates the name of the output file, including path and extension
   #'
   #' @value Returns a csv file containing summary statistics for each NHDv2 attribute variable
   
-  # Bind all columns representing unique NHDv2 attribute variables
-  attr_data_df <- attr_data %>%
-    Reduce(full_join,.) %>%
-    # hide messages that data frames are being joined by column 'PRMS_segid'
-    suppressMessages()
-  
+
   # Define function to summarize the number of NA's in numeric vector x
-  num_NA <- function(x){
+  # The argument `na.rm = TRUE` is included so that the `summarize_at` line will run in the code
+  # chunk below that generates `attr_summary`; it does not remove NA values in `num_NA()`
+  num_NA <- function(x,na.rm=TRUE){
     sum(is.na(x))
```

Comment:
Yes, I did! I found that I could either omit `na.rm=TRUE` from all of the pre-loaded summary functions, or, to include `num_NA()` in that same list, it needed the `na.rm` argument to match the format of the others. This isn't pretty, and I'm open to other suggestions 😃 

I also wondered whether the `num_NA` column as I have it is very useful. In our attribute table, each row represents a PRMS segment. An attribute will only equal `NA` for a PRMS segment if _all_ of the attribute values for the contributing NHD catchments are NA (or were -9999). Presumably these would be PRMS segments that only have one contributing COMID, because I remove NA's when applying the aggregation functions with `process_catchment_nhdv2_attr()`. I hope that makes sense!

Reply:
Thanks for clarifying what NAs indicate. Would it be difficult to also have a column for the total NAs (-9999) across all NHD catchments for each PRMS segment? That can help us make a decision about including a variable or not for model predictions. If many NHD catchments do not have data, then it could be that the dataset is not spatially available / reliable

Reply:
I agree that it'd be useful to know the number of NA's/(-9999) across the contributing NHD catchments. I think the reason I didn't do this before is because we'd need to add an extra column (e.g. '{col}_numberNA' or '{col}_proportionNA') for _each_ attribute variable, and so the returned target would be a mix of diagnostic information and actual values for each variable. So for example, the catchment attribute table for the STATSGO_HYDRGRP data might look like:
```

dat_proc_aggregated %>% head() %>% as.data.frame()
  PRMS_segid AREASQKM_PRMS CAT_HGA_area_wtd CAT_HGAC_area_wtd CAT_HGAD_area_wtd CAT_HGB_area_wtd CAT_HGBC_area_wtd CAT_HGBD_area_wtd CAT_HGC_area_wtd
1        1_1       65.0169         0.000000                 0         0.0000000         0.000000                 0                 0         90.41000
2       10_1        5.4666         8.470000                 0         0.0000000         4.390000                 0                 0         81.88000
3       11_1        5.9391         0.000000                 0         0.0000000         0.000000                 0                 0         96.00000
4      111_1       41.6439         1.360000                 0         0.0000000         0.700000                 0                 0         92.74000
5      112_1      270.7272         5.699743                 0         0.4252225         2.868315                 0                 0         85.21747
6      113_1       21.1104         4.679534                 0         0.0000000         2.424163                 0                 0         88.66857
  CAT_HGCD_area_wtd CAT_HGD_area_wtd CAT_HGVAR_area_wtd num_NHDv2cats CAT_HGA_num_NA CAT_HGAC_num_NA CAT_HGAD_num_NA CAT_HGB_num_NA CAT_HGBC_num_NA
1          7.750000        1.8400000                  0             1              0               0               0              0               0
2          3.840000        1.4100000                  0             1              0               0               0              0               0
3          4.000000        0.0000000                  0             1              0               0               0              0               0
4          4.000000        1.2000000                  0             1              0               0               0              0               0
5          3.151887        2.6372861                  0            24              0               0               0              0               0
6          3.444865        0.7828671                  0             3              0               0               0              0               0
  CAT_HGBD_num_NA CAT_HGC_num_NA CAT_HGCD_num_NA CAT_HGD_num_NA CAT_HGVAR_num_NA
1               0              0               0              0                0
2               0              0               0              0                0
3               0              0               0              0                0
4               0              0               0              0                0
5               0              0               0              0                0
6               0              0               0              0                0
> 
```

The summary table created in phase 3 is aggregated across the network (and all 459 PRMS segments). Is it useful to have summary columns that show the range in the number of NA's within the PRMS segments? Like this:

```
# A tibble: 11 x 7
   name                 Min   Mean   Max     Sd numNA_min numNA_max
   <chr>              <dbl>  <dbl> <dbl>  <dbl>     <int>     <int>
 1 AREASQKM_PRMS      0.001 71.0   539.  80.0          NA        NA
 2 CAT_HGA_area_wtd   0      6.54   60    8.85          0        40
 3 CAT_HGAC_area_wtd  0      0       0    0             0        40
 4 CAT_HGAD_area_wtd  0      0.235  10.4  0.983         0        40
 5 CAT_HGB_area_wtd   0     29.3    82   25.0           0        40
 6 CAT_HGBC_area_wtd  0      0.256  10    1.2           0        40
 7 CAT_HGBD_area_wtd  0      4.50   31    6.15          0        40
 8 CAT_HGC_area_wtd   0     41.6   100   30.5           0        40
 9 CAT_HGCD_area_wtd  0      3.93   49.2  6.14          0        40
10 CAT_HGD_area_wtd   0     12.4    80   15.7           0        40
11 CAT_HGVAR_area_wtd 0      0.499  18    2.31          0        40
```

If not, I could create a different target that logs the NA's for each segment and variable. Or do you have other suggestions?

Reply:
I like seeing the NA min and max. I didn't realize that some HRUs had >40 NHD catchments! 

I think it will help for feature selection to have the ` target that logs the NA's for each segment and variable`. I think what would be most helpful is to report the proportion of total NHD area that is not NA for each PRMS segment, and then we can decide to omit a variable if at least X% of the catchment is not NA. Does that make sense?

Reply:
I changed the output of `process_catchment_nhdv2_attr()` so that the aggregated data gets returned in addition to a table that summarizes the NA diagnostics. The NA diagnostics for each segment are collected and saved in a new target `p3_nhdv2_attr_missing_data_csv` that contains columns for the number of contributing NHDv2 catchments for each PRMS segment, the number of those catchments with missing/flagged data (`_num_NA`) and the proportion of the catchment area with missing/flagged data based on the contributing NHDv2 catchments (`_propArea_NA`). Only the attribute variables that have -9999 or NA within the study extent (n=13) will get returned in this diagnostics table.

Finally, since the network-scale NA diagnostics are less useful, I've omitted the `num_NA` column from the summary table `p3_nhdv2_attr_summary_csv`. Hopefully these edits will help us further evaluate the attribute variables and their spatial coverage - let me know what you think.

Reply:
Looks great! The new missing data table is very helpful. It's good to know that only 13 variables are affected by NAs. Some of the NHD areas must be smaller than your rounding tolerance (propAREA_NA shows as 0 and num_NA shows as 1), but I think that's okay because we wouldn't care about missing such a small proportion of the area.

Reply:
> Some of the NHD areas must be smaller than your rounding tolerance (propAREA_NA shows as 0 and num_NA shows as 1)

Yeah, that's right. As an example, here's the data for one variable `CAT_HGAC` for `PRMS_segid == "140_1"`. There's a flagged value for one COMID among the NHD catchments that make up the PRMS catchment. The proportion of area represented by that one COMID is ~1.5e-05, but it shows as zero since I round those proportion area estimates to 4 digits. 

![example](https://user-images.githubusercontent.com/8785034/156230421-f82edd3a-2df5-4675-918d-14f8d59efebb.png)

Since `CAT_HGAC` is a STATSGO variable, we change that -9999 to 0 before calculating the area-weighted mean.

Reply:
Thanks, sounds good!